ggml.c

```diff
@@ -1262,6 +1261,47 @@ inline static void ggml_vec_dot_f32(const int n, float * restrict s, const float
     *s = sumf;
 }
 
+#if __AVX512F__ && QK == 32
+static inline __m512 dot_q4_0_oneblock_avx512(
+    __m512 acc,
+    const uint8_t * pd0,
+    const uint8_t * pd1,
+    const uint8_t * pb0,
+    const uint8_t * pb1,
+    size_t bs,
+    int i
+) {
+    const float * d0_0 = (const float *) (pd0 + i*bs);
+    const float * d1_0 = (const float *) (pd1 + i*bs);
```

Comment:
Pointers galore ;-)

I think this can made quite a bit simpler, but maybe that's for a separate PR.

The block size `bs` is fixed by `QK`, and you shouldn't really need separate pointers for the scale and the q4 values.

I'm thinking something like this:
```c
typedef struct {
  float scale;
  uin8_t q4[QK/2];
} q4_block;
```
Then just pass around two pointers of `*q4_block` for x and y.

Reply:
I tried to leave it as close as possible to the way it was before, just splitting it out to facilitate the manual unrolling with two accumulators to help with instruction pipelining and floating point numerical precision.

The start of the `ggml_vec_dot_q4_0` function defines those two different points for scales and quantized values separately, so I just keep using those instead of dong my own pointer arethmetic.

That being said, I'm not attached to the particulars of the arguments or layout or whatever.  If you or some other maintainer has some preference for how it's handled, feel free to just push to this branch with changes or let me know what you want me to change.

Reply:
That's ok for now. I don't have AVX512 hardware though, so maybe someone else can have a look at this before merging it.